design_drafts/joint_interfaces_and_transmissions.md

```diff
@@ -0,0 +1,145 @@
+# Joint/actuator interfaces and transmission parsing
+
+In ROS Melodic and previous distributions, `ros_control` operates with statically-typed joint handles as the primary mean of communication within components of the framework. 
+These handles are stored in joint-name to handle maps and are made available to controllers inheriting (or using other means to acquire) the relevant controller interface class e.g.
+```
+  class DiffDriveController : 
+    public controller_interface::Controller<hardware_interface::VelocityJointInterface>
+  { ... };
+```
+
+Commands on joint interfaces are transformed into commands for actuator interfaces using transmissions.
+An example of such a transmission is simple reducer which applies a multiplier between commands and readings.
+Users of `ros_control` can extend the framework by providing their own transmissions as all transmissions are loaded as plugins.
+When defining a robot setup, joints and actuators are connected via transmissions declared through the URDF along with the required interfaces.
+At startup, the required interfaces for transmissions are validated by the `TransmissionLoader`.
+Transmissions use a [common interface](https://github.com/ros-controls/ros_control/blob/melodic-devel/transmission_interface/include/transmission_interface/transmission.h) to propagate values.
+This however already shows a shortcomings of this design as a whole as with a fixed set of interfaces, a fixed set of functions were created, most of them defined empty, for example:
+```
+  virtual void jointToActuatorEffort(const JointData&    jnt_data,
+                                           ActuatorData& act_data) = 0;
+```
+All transmissions inheriting this interface has to define empty functions for those it does not operate on. Additionally, this set of functions is not extendable in a clean way, without incurring additional interface inheritance (and a general mess...) on the derived classes.
+ 
+## Scope:
+This design document proposes changes to 
+1) joint and actuator interfaces
+2) transmission parsing
+3) controller interfaces (as a consequence of 1)
+
+and ties in with the use of the [flexible joint states message](https://github.com/ros-controls/roadmap/blob/master/design_drafts/flexible_joint_states_msg.m) 
+4) in the context of replacing `ActuatorData` and `JointData` from `ros_control` on ROS Melodic in how transmission loading and runtime is handled
+5) the `joint_state_controller` which depends on the "read-only" `JointStateHandle` class to report position, velocity and effort values.
```

Comment:
The 4) and 5) points are broken in the resulting list of the markdown document.

Reply:
fixed